File: native/haskell/HSNMask.hs
{-# LANGUAGE ForeignFunctionInterface #-}
-- Optional Haskell sidecar kernels. Built with GHC -shared when present.
-- Never required on the Minecraft frame path (see ARCHITECTURE.md).

module HSNMask where

import Data.Bits ((.&.), (.|.), shiftL)
import Foreign.C.Types
import Foreign.Ptr
import Foreign.Storable

dropF64 :: CDouble -> CDouble -> CUChar
dropF64 value limitSq = if value > limitSq then 1 else 0

fillMask :: Ptr CDouble -> CDouble -> Ptr CUChar -> CSize -> IO ()
fillMask input limitSq out n = go 0
  where
    go i
      | i >= n = return ()
      | otherwise = do
          v <- peekElemOff input (fromIntegral i)
          pokeElemOff out (fromIntegral i) (dropF64 v limitSq)
          go (i + 1)

distSq :: Ptr CDouble -> Ptr CDouble -> Ptr CDouble
       -> CDouble -> CDouble -> CDouble
       -> Ptr CDouble -> CSize -> IO ()
distSq xs ys zs ox oy oz out n = go 0
  where
    go i
      | i >= n = return ()
      | otherwise = do
          x <- peekElemOff xs (fromIntegral i)
          y <- peekElemOff ys (fromIntegral i)
          z <- peekElemOff zs (fromIntegral i)
          let dx = x - ox
              dy = y - oy
              dz = z - oz
          pokeElemOff out (fromIntegral i) (dx * dx + dy * dy + dz * dz)
          go (i + 1)

lodBand :: Ptr CDouble -> CDouble -> CDouble -> Ptr CUChar -> CSize -> IO ()
lodBand dist nearSq midSq out n = go 0
  where
    go i
      | i >= n = return ()
      | otherwise = do
          d <- peekElemOff dist (fromIntegral i)
          let band | d < nearSq = 0
                   | d < midSq  = 1
                   | otherwise  = 2
          pokeElemOff out (fromIntegral i) band
          go (i + 1)

rsqrtF64 :: Ptr CDouble -> Ptr CDouble -> CSize -> IO ()
rsqrtF64 input out n = go 0
  where
    go i
      | i >= n = return ()
      | otherwise = do
          v <- peekElemOff input (fromIntegral i)
          let y = if v > 0 then 1.0 / sqrt (realToFrac v :: Double) else 0
          pokeElemOff out (fromIntegral i) (realToFrac y)
          go (i + 1)

frustumAabb :: Ptr CFloat -> Ptr CFloat -> Ptr CUChar -> CSize -> IO ()
frustumAabb planes aabb out n
  | n == 0 = return ()
  | otherwise = go 0
  where
    outside a b c d minx miny minz maxx maxy maxz =
      let px = if a >= 0 then maxx else minx
          py = if b >= 0 then maxy else miny
          pz = if c >= 0 then maxz else minz
      in a * px + b * py + c * pz + d < 0
    go i
      | i >= n = return ()
      | otherwise = do
          let base = fromIntegral i * 6
          minx <- peekElemOff aabb base
          miny <- peekElemOff aabb (base + 1)
          minz <- peekElemOff aabb (base + 2)
          maxx <- peekElemOff aabb (base + 3)
          maxy <- peekElemOff aabb (base + 4)
          maxz <- peekElemOff aabb (base + 5)
          dropBox <- checkPlane 0 False minx miny minz maxx maxy maxz
          pokeElemOff out (fromIntegral i) (if dropBox then 1 else 0)
          go (i + 1)
    checkPlane p acc minx miny minz maxx maxy maxz
      | acc = return True
      | p >= 6 = return False
      | otherwise = do
          let o = p * 4
          a <- peekElemOff planes o
          b <- peekElemOff planes (o + 1)
          c <- peekElemOff planes (o + 2)
          d <- peekElemOff planes (o + 3)
          checkPlane (p + 1) (outside a b c d minx miny minz maxx maxy maxz)
                     minx miny minz maxx maxy maxz

part1by2 :: CUInt -> CUInt
part1by2 n0 =
  let n1 = n0 .&. 0x3ff
      n2 = (n1 .|. shiftL n1 16) .&. 0x030000FF
      n3 = (n2 .|. shiftL n2 8)  .&. 0x0300F00F
      n4 = (n3 .|. shiftL n3 4)  .&. 0x030C30C3
      n5 = (n4 .|. shiftL n4 2)  .&. 0x09249249
  in n5

morton3 :: Ptr CInt -> Ptr CInt -> Ptr CInt -> Ptr CUInt -> CSize -> IO ()
morton3 xs ys zs out n = go 0
  where
    go i
      | i >= n = return ()
      | otherwise = do
          x <- peekElemOff xs (fromIntegral i)
          y <- peekElemOff ys (fromIntegral i)
          z <- peekElemOff zs (fromIntegral i)
          let xi = part1by2 (fromIntegral x)
              yi = part1by2 (fromIntegral y)
              zi = part1by2 (fromIntegral z)
          pokeElemOff out (fromIntegral i) (xi .|. shiftL yi 1 .|. shiftL zi 2)
          go (i + 1)

keepU8 :: Ptr CUChar -> CUChar -> Ptr CUChar -> CSize -> IO ()
keepU8 rnd thresh out n = go 0
  where
    go i
      | i >= n = return ()
      | otherwise = do
          v <- peekElemOff rnd (fromIntegral i)
          pokeElemOff out (fromIntegral i) (if v < thresh then 1 else 0)
          go (i + 1)

hsn_hs_cull_f64 :: Ptr CDouble -> CDouble -> Ptr CUChar -> CSize -> IO ()
hsn_hs_cull_f64 input limitSq out n
  | n == 0 = return ()
  | otherwise = fillMask input limitSq out n

hsn_hs_dist_sq :: Ptr CDouble -> Ptr CDouble -> Ptr CDouble
               -> CDouble -> CDouble -> CDouble
               -> Ptr CDouble -> CSize -> IO ()
hsn_hs_dist_sq xs ys zs ox oy oz out n
  | n == 0 = return ()
  | otherwise = distSq xs ys zs ox oy oz out n

hsn_hs_lod_band :: Ptr CDouble -> CDouble -> CDouble -> Ptr CUChar -> CSize -> IO ()
hsn_hs_lod_band = lodBand

hsn_hs_rsqrt_f64 :: Ptr CDouble -> Ptr CDouble -> CSize -> IO ()
hsn_hs_rsqrt_f64 = rsqrtF64

hsn_hs_frustum_aabb :: Ptr CFloat -> Ptr CFloat -> Ptr CUChar -> CSize -> IO ()
hsn_hs_frustum_aabb = frustumAabb

hsn_hs_morton3 :: Ptr CInt -> Ptr CInt -> Ptr CInt -> Ptr CUInt -> CSize -> IO ()
hsn_hs_morton3 = morton3

hsn_hs_keep_u8 :: Ptr CUChar -> CUChar -> Ptr CUChar -> CSize -> IO ()
hsn_hs_keep_u8 = keepU8

hsn_hs_abi :: IO CInt
hsn_hs_abi = return 0x48534B4C -- "HSKL"

foreign export ccall hsn_hs_cull_f64 :: Ptr CDouble -> CDouble -> Ptr CUChar -> CSize -> IO ()
foreign export ccall hsn_hs_dist_sq :: Ptr CDouble -> Ptr CDouble -> Ptr CDouble -> CDouble -> CDouble -> CDouble -> Ptr CDouble -> CSize -> IO ()
foreign export ccall hsn_hs_lod_band :: Ptr CDouble -> CDouble -> CDouble -> Ptr CUChar -> CSize -> IO ()
foreign export ccall hsn_hs_rsqrt_f64 :: Ptr CDouble -> Ptr CDouble -> CSize -> IO ()
foreign export ccall hsn_hs_frustum_aabb :: Ptr CFloat -> Ptr CFloat -> Ptr CUChar -> CSize -> IO ()
foreign export ccall hsn_hs_morton3 :: Ptr CInt -> Ptr CInt -> Ptr CInt -> Ptr CUInt -> CSize -> IO ()
foreign export ccall hsn_hs_keep_u8 :: Ptr CUChar -> CUChar -> Ptr CUChar -> CSize -> IO ()
foreign export ccall hsn_hs_abi :: IO CInt

File: native/swift/HSNMask.swift
#if os(Linux)
import Glibc
#else
import Darwin
#endif

/// Extra Swift kernels (mask / LOD / Morton). Optional sidecar.

@inline(__always)
func hsnPart1By2(_ raw: UInt32) -> UInt32 {
    var n = raw & 0x0000_03ff
    n = (n | (n << 16)) & 0x0300_00FF
    n = (n | (n << 8)) & 0x0300_F00F
    n = (n | (n << 4)) & 0x030C_30C3
    n = (n | (n << 2)) & 0x0924_9249
    return n
}

@_cdecl("hsn_swift_lod_band")
public func hsnSwiftLodBand(_ distSq: UnsafePointer<Double>?,
                            _ nearSq: Double,
                            _ midSq: Double,
                            _ out: UnsafeMutablePointer<UInt8>?,
                            _ n: UInt64) {
    guard let distSq, let out, n > 0 else { return }
    var i: UInt64 = 0
    while i < n {
        let d = distSq[Int(i)]
        if d < nearSq {
            out[Int(i)] = 0
        } else if d < midSq {
            out[Int(i)] = 1
        } else {
            out[Int(i)] = 2
        }
        i += 1
    }
}

@_cdecl("hsn_swift_rsqrt_f64")
public func hsnSwiftRsqrt(_ input: UnsafePointer<Double>?,
                          _ out: UnsafeMutablePointer<Double>?,
                          _ n: UInt64) {
    guard let input, let out, n > 0 else { return }
    var i: UInt64 = 0
    while i < n {
        let v = input[Int(i)]
        if v > 0 {
            var y = 1.0 / v.squareRoot()
            y = y * (1.5 - 0.5 * v * y * y)
            out[Int(i)] = y
        } else {
            out[Int(i)] = 0
        }
        i += 1
    }
}

@_cdecl("hsn_swift_morton3")
public func hsnSwiftMorton3(_ x: UnsafePointer<Int32>?,
                            _ y: UnsafePointer<Int32>?,
                            _ z: UnsafePointer<Int32>?,
                            _ out: UnsafeMutablePointer<UInt32>?,
                            _ n: UInt64) {
    guard let x, let y, let z, let out, n > 0 else { return }
    var i: UInt64 = 0
    while i < n {
        let idx = Int(i)
        let xi = hsnPart1By2(UInt32(bitPattern: x[idx]) & 0x3ff)
        let yi = hsnPart1By2(UInt32(bitPattern: y[idx]) & 0x3ff)
        let zi = hsnPart1By2(UInt32(bitPattern: z[idx]) & 0x3ff)
        out[idx] = xi | (yi << 1) | (zi << 2)
        i += 1
    }
}

@_cdecl("hsn_swift_keep_u8")
public func hsnSwiftKeep(_ rnd: UnsafePointer<UInt8>?,
                         _ thresh: UInt8,
                         _ out: UnsafeMutablePointer<UInt8>?,
                         _ n: UInt64) {
    guard let rnd, let out, n > 0 else { return }
    var i: UInt64 = 0
    while i < n {
        out[Int(i)] = rnd[Int(i)] < thresh ? 1 : 0
        i += 1
    }
}

@_cdecl("hsn_swift_frustum_aabb")
public func hsnSwiftFrustumAabb(_ planes: UnsafePointer<Float>?,
                                _ aabb: UnsafePointer<Float>?,
                                _ out: UnsafeMutablePointer<UInt8>?,
                                _ n: UInt64) {
    guard let planes, let aabb, let out, n > 0 else { return }
    var i: UInt64 = 0
    while i < n {
        let b = Int(i) * 6
        let minx = aabb[b], miny = aabb[b + 1], minz = aabb[b + 2]
        let maxx = aabb[b + 3], maxy = aabb[b + 4], maxz = aabb[b + 5]
        var drop: UInt8 = 0
        var p = 0
        while p < 6 {
            let o = p * 4
            let a = planes[o], b0 = planes[o + 1], c = planes[o + 2], d = planes[o + 3]
            let px = a >= 0 ? maxx : minx
            let py = b0 >= 0 ? maxy : miny
            let pz = c >= 0 ? maxz : minz
            if a * px + b0 * py + c * pz + d < 0 {
                drop = 1
                break
            }
            p += 1
        }
        out[Int(i)] = drop
        i += 1
    }
}

@_cdecl("hsn_swift_mask_abi")
public func hsnSwiftMaskAbi() -> Int32 {
    return 0x53575446 // "SWTF"
}
